src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerActor.scala

```diff
@@ -266,13 +268,30 @@ private[impl] class GroupManagerActor(
 
     val dynamicApps: Set[AppDefinition] =
       to.transitiveApps.map {
-        // assign values for service ports that the user has left "blank" (set to zero)
-        case app: AppDefinition if app.hasDynamicServicePorts => assignPorts(app)
-        case app: AppDefinition =>
-          // Always set the ports to service ports, even if we do not have dynamic ports in our port mappings
+        case app: AppDefinition if app.servicePorts.nonEmpty =>
+          synchronized {
+            app.servicePorts.withFilter(_ != AppDefinition.RandomPortValue).foreach { port => taken += port }
+          }
+          assignPorts(app)
+        case app: AppDefinition if app.portDefinitions.exists(_.port == AppDefinition.RandomPortValue) =>
```

Comment:
`portDefinitions` were supposed to be a superset of `ports`. I think that we shouldn't make breaking changes of the interface, without having clearly communicated a deprecation plan.

Until recently, our docs used to say that the user was responsibly for ensuring the uniqueness of the service ports, and there was no validation.

Then the docs were updated to say that Marathon enforces the uniqueness, and a broken validation was added.

I recently fixed the validation. But I'd be ok with going back to not enforcing the uniqueness, as long as that is properly documented and maybe even checked with the community.